R/src.R

```diff
@@ -6,6 +6,98 @@
+#' @export
+#' @family parsers
+understand_parse_dependencies <- function(project_path,language,output_dir="../tmp/", parse_type){
+  # Before running, check if parse_type is correct
+  validInput <- function(input) {
+    is.character(input) && length(input) == 1 && input %in% c("file", "class")
+  }
+  if (!validInput(parse_type)) {
+    stop("Error: Invalid parse_type provided. Please input either \"file\" or \"class\"")
+  }
+
+  # Use Understand to parse the code folder.
+  # Create the variables used in command lines
+  project_path <- paste0("\"", project_path, "\"")
+  db_dir <- paste0(output_dir, "/Understand.und")
+  xml_dir <- paste0(db_dir, "/", parse_type, "Dependencies.xml")
+  command <- "und"
+  args <- c("create", "-db", db_dir, "-languages", language)
+
+  # Generate the XML file
+  system2(command, args)
+  args <- c("-db", db_dir, "add", project_path)
+  system2(command, args)
+  args <- c("analyze", db_dir)
+  system2(command, args)
+  args <- c("export", "-dependencies", parse_type, "cytoscape", xml_dir, db_dir)
+  system2(command, args)
+
+  # Parse the XML file
+  xml_data <- xmlParse(xml_dir)
+  xml_nodes <- xmlRoot(xml_data)
+  xml_nodes <- xmlChildren(xml_nodes)
+
+  # Helper function to search for an attribute
+  findAtt <- function(search_nodes, att_name) {
+    found_att <- NA
+    for (att in search_nodes) {
+      if (xmlGetAttr(att, "name") == att_name) {
+        found_att <- xmlGetAttr(att, "value")
+        break
+      }
+    }
+    return(found_att)
+  }
+
+  # From child nodes- filter for those with name "node"
+  node_elements <- lapply(xml_nodes, function(child) {
+    if (xmlName(child) == "node") {
+      # Extract the id
+      id <- xmlGetAttr(child, "id")
+      # Find the node.label attribute
+      att_nodes <- xmlChildren(child)
+      node_label <- findAtt(att_nodes, "node.label")
+      long_name <- findAtt(att_nodes, "longName")
+      return(data.table(node_label = node_label, id = id, long_name = long_name))
+    } else {
+      return(NULL)
+    }
+  })
+
+  # Remove NULLs and combine the results into a data frame
+  node_list <- do.call(rbind, node_elements[!sapply(node_elements, is.null)])
+
+  # From child nodes- filter for those with name "edge"
+  edge_elements <- lapply(xml_nodes, function(child) {
+    if (xmlName(child) == "edge") {
+      # Extract the id_from and id_to
+      id_from <- xmlGetAttr(child, "source")
+      id_to <- xmlGetAttr(child, "target")
+      # Find the dependency kind attribute
+      att_nodes <- xmlChildren(child)
+      dependency_kind <- findAtt(att_nodes, "dependency kind")
+      dependency_kind <- unlist(strsplit(dependency_kind, ",\\s*"))
+      return(data.table(id_from = id_from, id_to = id_to, dependency_kind = dependency_kind))
+    } else {
+      return(NULL)
+    }
+  })
+
+  # Remove NULLs and combine the results into a data frame
+  edge_list <- do.call(rbind, edge_elements[!sapply(edge_elements, is.null)])
```

Comment:
there is a data.table function that binds multiple lists into a table. Use that instead.